[draft][pythonic config][docs] Introduce intro to Resources doc utilizing Pythonic resources (#12260)
dagster-io/dagster: docs/content/guides/dagster/pythonic-resources.mdx

Resources are objects that are shared across the implementations of multiple [software-defined assets](/concepts/assets/software-defined-assets) and [ops](/concepts/ops-jobs-graphs/ops) and that can be plugged in after defining those ops and assets.

Resources typically model external components that assets and ops interact with. For example, a resource might be a connection to a data warehouse like Snowflake or a service like Slack.

So, why use resources?

- **Plug in different implementations in different environments** - If you have a heavy external dependency that you want to use in production, but avoid using in testing, you can accomplish this by providing different resources in each environment. Check out [Separating Business Logic from Environments](/concepts/testing#separating-business-logic-from-environments) for more info about this capability.
- **Share configuration across multiple ops or assets** - Resources are configurable and shared, so you can supply configuration in one place instead of configuring the ops and assets individually.

> - **Share implementations across multiple ops or assets** - When multiple ops access the same external services, resources provide a standard way to structure your code to share the implementations.

Reviewer: We should add a top bullet that speaks to new Dagit features that display resources and their configuration. I think this observability is now the top reason to use them

---

## Using resources

### Getting started using resources

> A resource can be any Python object which you would like to share between software-defined assets or ops. Assets and ops specify resource dependencies by annotating the resource as a parameter to the asset or op function. The parameter type should be wrapped in a <PyObject object="Resource"/> annotation to indicate that it is a resource and not an upstream asset or op input.

Reviewer: `The parameter type should be wrapped in a <PyObject object="Resource"/> annotation to indicate that it is a resource and not an upstream asset or op input.`

This is going to present a challenge for sure. I consider this an advanced capability but you are right that users will probably encounter it early in their journey. Still I would much prefer to try to introduce the `Resource` annotation later and instead assume they will annotate with their resource definition

Author: moved out of critical path to an 'advanced usage' section

To provide resource values to your assets and ops, attach them to your <PyObject object="Definitions" /> call. These resources are automatically passed to the function at runtime.

<TabGroup persistentKey="assetsorops">
<TabItem name="Using software-defined-assets">

#### Using software-defined assets

> Here, we define a simple string resource representing a URL which we use in a software-defined asset. We specify a value for the URL in the <PyObject object="Definitions" /> call.

Reviewer: The first example should be something more typically, like a database connection. Whatever it is it should be a class.

Reviewer: In fact, I don't think we should encourage the usage of resources like this at all.

Author: moved out of critical path to an 'advanced usage' section

from dagster import asset, Definitions
from dagster._core.definitions.resource_annotation import Resource

@asset
def data_from_url(data_url: Resource[str]) -> Dict[str, Any]:
    return requests.get(data_url).json()

defs = Definitions(
    assets=[data_from_url],
    resources={"data_url": "https://dagster.io"},
)
```

</TabItem>
<TabItem name="Using ops and jobs">

#### Using ops

Here, we define a simple string resource representing a URL which we use in an op. We specify a value for the URL in the <PyObject object="Definitions" /> call.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_ops_defs endbefore=end_new_resources_ops_defs
from dagster import op, Definitions, job
from dagster._core.definitions.resource_annotation import Resource

@op
def print_data_from_resource(data_url: Resource[str]):
    print(requests.get(data_url).json())

@job
def print_data_from_url_job():
    print_data_from_resource()

defs = Definitions(
    jobs=[print_data_from_url_job],
    resources={"data_url": "https://dagster.io"},
)
```

</TabItem>
</TabGroup>

### Configurable resources

> In many cases, resources need to be [configurable](/concepts/configuration/config-schema). For example, a resource representing a connection to an external service may need a configurable username and password. Configurable resources can be defined by extending the <PyObject object="ConfigurableResource"/> class. The attributes on the class are used to define the configuration schema, which lets you customize behavior through run configuration.

Reviewer: We should answer the question, why "configuration" rather than just plain ole python parameter passing here

We can also define methods on the resource which rely on the config values.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_configurable_defs_ops endbefore=end_new_resources_configurable_defs_ops
from dagster import Definitions, job, op
from dagster._config.structured_config import ConfigurableResource

class MyConnectionResource(ConfigurableResource):
    username: str
    password: str

    def request(self, endpoint: str) -> Response:
        return requests.get(
            f"https://my-api.com/{endpoint}",
            auth=(self.username, self.password),
        )

# Since MyConnectionResource extends ConfigurableResource, we don't
# need to wrap it in a Resource[] annotation.
@op
def update_service(my_conn: MyConnectionResource):
    my_conn.request("/update")

@job
def update_service_job():
    update_service()

defs = Definitions(
    jobs=[update_service_job],
    resources={
        "my_conn": MyConnectionResource(username="my_user", password="my_password"),
    },
)
```

</TabItem>
</TabGroup>

---

> ## Using environment variables with resources

Reviewer: Note somewhere that we have special features in Cloud to support in-UI env variables.

Resources can be configured using environment variables, which is useful for secrets or other environment-specific configuration.

To use environment variables, pass an <PyObject object="EnvVar" /> when constructing your resource.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_env_vars endbefore=end_new_resources_env_vars
from dagster._config.field_utils import EnvVar
from dagster import Definitions
from dagster._config.structured_config import ConfigurableResource

class CredentialsResource(ConfigurableResource):
    username: str
    password: str

defs = Definitions(
    assets=...,
    resources={
        "credentials": CredentialsResource(
            username=EnvVar("MY_USERNAME"),
            password=EnvVar("MY_PASSWORD"),
        )
    },
)
```

---

## Configuring resources at runtime

> In some cases, you may want to specify configuration for a resource at runtime, in the launchpad. For example, you may want to specify a different target table in a database resource for each run.

Reviewer: Other critical use case is RunRequest in schedules and sensors

To do this, you can use the `configure_at_launch()` method to defer the construction of a configurable resource until runtime.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_runtime endbefore=end_new_resource_runtime
class DatabaseResource(ConfigurableResource):
    table: str

    def read(self):
        ...

@asset
def data_from_database(db_conn: DatabaseResource):
    return db_conn.read()

defs = Definitions(
    assets=...,
    resources={"db_conn": DatabaseResource.configure_at_launch()},
)
```

---

## Resources which depend on other resources

> In some situations, you may want to define a resource which depends on other resources. For example, a resource representing a connection to an external service may rely on a resource representing credentials for that service.

Reviewer: I think we need to flesh out the motivation here.

Author: added some more explanation

In this case, you can list that nested resource as an attribute of your resource class.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_nesting endbefore=end_new_resources_nesting
from dagster import Definitions
from dagster._config.structured_config import ConfigurableResource

class CredentialsResource(ConfigurableResource):
    username: str
    password: str

class FileStoreBucket(ConfigurableResource):
    credentials: CredentialsResource
    region: str

    def write(self, data: str):
        get_filestore_client(
            username=self.credentials.username,
            password=self.credentials.password,
            region=self.region,
        ).write(data)

credentials = CredentialsResource(username="my_user", password="my_password")
defs = Definitions(
    assets=...,
    resources={
        "bucket": FileStoreBucket(
            credentials=credentials,
            region="us-east-1",
        ),

>     },

Reviewer: We should decide on best practice in terms of whether or not we encourage people to a set a top-level resource for something like credentials here, even if we do support it in this style.

)
```

If we instead would like the configuration for our credentials to be provided at runtime, we can use the `configure_at_launch()` method to defer the construction of the `CredentialsResource` until runtime.

> Because `credentials` requires runtime configuration through the launchpad, it must also be passed to the `Definitions` object, so that configuration can be provided at runtime. Nested resources only need to be passed to the `Definitions` object if they require runtime configuration.

Reviewer: This is good example for motivation behind resource-to-resource specifies: canonicalization of runtime config to a single location.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_testing endbefore=end_new_resource_testing
from dagster._config.structured_config import ConfigurableResource

class MyResource(ConfigurableResource):
    def get_value(self) -> str:
        return "foo"

def test_my_resource():
    assert MyResource().get_value() == "foo"
```

If your resource requires other resources, then you can pass them as constructor arguments.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_testing_with_context endbefore=end_new_resource_testing_with_context
from dagster._config.structured_config import ConfigurableResource

class StringHolderResource(ConfigurableResource):
    value: str

class MyResourceRequiresAnother(ConfigurableResource):
    foo: StringHolderResource
    bar: str

def test_my_resource_with_context():
    resource = MyResourceRequiresAnother(foo=StringHolderResource("foo"), bar="bar")
    assert resource.foo.value == "foo"
    assert resource.bar == "bar"

> ```

Reviewer: Let's add a section describe typing and how we accept different types at the constructor than the output. E.g. you can pass `EnvVar` in but you always get `str` out.

Author: Added in an `implementation details` section

So, why use resources?

- **Plug in different implementations in different environments** - If you have a heavy external dependency that you want to use in production, but avoid using in testing, you can accomplish this by providing different resources in each environment. Check out [Separating Business Logic from Environments](/concepts/testing#separating-business-logic-from-environments) for more info about this capability.
- **Surface configuration in the UI** - Resources and their configuration are surfaced in the Dagster UI, making it easy to see where your resources are used and how they are configured.
- **Share configuration across multiple ops or assets** - Resources are configurable and shared, so you can supply configuration in one place instead of configuring the ops and assets individually.
- **Share implementations across multiple ops or assets** - When multiple ops access the same external services, resources provide a standard way to structure your code to share the implementations.

---

## Using Pythonic resources

> ### Getting started using resources

Reviewer 2: Nit - is `using resources` in this heading necessary? Given that the heading above it is `Using Pythonic resources`, it feels a bit redundant IMO. `Getting started` could be enough here.

Here, we define a subclass of <PyObject object="ConfigurableResource"/> representing a connection to an external service. We can configure the resource by constructing it in the <PyObject object="Definitions" /> call.

We can define methods on the resource class which depend on config values. These methods can be used by assets and ops.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_configurable_defs endbefore=end_new_resources_configurable_defs
from dagster import asset, Definitions
from dagster._config.structured_config import ConfigurableResource

class MyConnectionResource(ConfigurableResource):
    username: str
    password: str

    def request(self, endpoint: str) -> Response:
        return requests.get(
            f"https://my-api.com/{endpoint}",
            auth=(self.username, self.password),
        )

@asset
def data_from_service(my_conn: MyConnectionResource) -> Dict[str, Any]:
    return my_conn.request("/fetch_data").json()

defs = Definitions(
    assets=[data_from_service],
    resources={
        "my_conn": MyConnectionResource(username="my_user", password="my_password"),
    },
)
```

> </TabItem>

Reviewer: Here we should either (1) use a different example or (2) note that using a bare string for a password is a very bad idea. I bias towards (1) and then only introduce a password in the environment variable section.

Reviewer 2: +1

As the ops example is essentially the same thing, I assume this comment should extend to that example as well?

We can define methods on the resource class which depend on config values. These methods can be used by assets and ops.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_configurable_defs_ops endbefore=end_new_resources_configurable_defs_ops
from dagster import Definitions, job, op
from dagster._config.structured_config import ConfigurableResource

class MyConnectionResource(ConfigurableResource):
    username: str
    password: str

    def request(self, endpoint: str) -> Response:
        return requests.get(
            f"https://my-api.com/{endpoint}",
            auth=(self.username, self.password),
        )

@op
def update_service(my_conn: MyConnectionResource):
    my_conn.request("/update")

@job
def update_service_job():
    update_service()

defs = Definitions(
    jobs=[update_service_job],
    resources={
        "my_conn": MyConnectionResource(username="my_user", password="my_password"),
    },
)
```

</TabItem>
</TabGroup>

> ---

Reviewer 2: [1] Rules are only needed before H2 headings, so we can take this guy out

### Using environment variables with resources

Resources can be configured using environment variables, which is useful for secrets or other environment-specific configuration. If you're using Dagster Cloud, environment variables can be [configured directly in the UI](/dagster-cloud/developing-testing/environment-variables-and-secrets).

To use environment variables, pass an <PyObject object="EnvVar" /> when constructing your resource.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_env_vars endbefore=end_new_resources_env_vars
from dagster._config.field_utils import EnvVar
from dagster import Definitions
from dagster._config.structured_config import ConfigurableResource

class CredentialsResource(ConfigurableResource):
    username: str
    password: str

defs = Definitions(
    assets=...,
    resources={
        "credentials": CredentialsResource(
            username=EnvVar("MY_USERNAME"),
            password=EnvVar("MY_PASSWORD"),
        )
    },
)
```

> ---

Reviewer 2: [1]

### Configuring resources at runtime

In some cases, you may want to specify configuration for a resource at runtime, in the launchpad or in a <PyObject object="RunRequest" /> for a [schedule](/concepts/partitions-schedules-sensors/schedules) or [sensor](/concepts/partitions-schedules-sensors/sensors). For example, you may want a sensor-triggered run to specify a different target table in a database resource for each run.

You can use the `configure_at_launch()` method to defer the construction of a configurable resource until runtime.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_runtime endbefore=end_new_resource_runtime
class DatabaseResource(ConfigurableResource):
    table: str

    def read(self):
        ...

@asset
def data_from_database(db_conn: DatabaseResource):
    return db_conn.read()

defs = Definitions(
    assets=...,
    resources={"db_conn": DatabaseResource.configure_at_launch()},
)
```

> ---

Reviewer 2: [1]

> 

Reviewer: Recommend showing a RunRequest construction with this config

Reviewer 2: +1

If we want to have two examples here, we could put them into tabs.

Ex: `Title for the basic example` | `Using a RunRequest`

> ### Resources which depend on other resources

Reviewer 2: ```suggestion
### Resources that depend on other resources
```

> In some situations, you may want to define a resource which depends on other resources. This is useful for common configuration. For example, separate resources for a database and for a filestore may both depend on credentials for a particular cloud provider. Defining these credentials as a separate, nested resource allows you to specify configuration in a single place.

Reviewer: Additional benefit is that this can assist in the testability of resources themselves.

> In some situations, you may want to define a resource which depends on other resources. This is useful for common configuration. For example, separate resources for a database and for a filestore may both depend on credentials for a particular cloud provider. Defining these credentials as a separate, nested resource allows you to specify configuration in a single place.

Reviewer 2: ```suggestion
In some situations, you may want to define a resource that depends on other resources. This is useful for common configuration. For example, separate resources for a database and for a filestore may both depend on credentials for a particular cloud provider. Defining these credentials as a separate, nested resource allows you to specify configuration in a single place.
```

> In this case, you can list that nested resource as an attribute of your resource class.

Reviewer 2: ```suggestion
In this case, you can list the nested resource as an attribute of your resource class.
```

This feels like a nit (sorry!) but using `that` here makes me think we're referring to a specific nested resource when we're actually speaking about it in general

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_nesting endbefore=end_new_resources_nesting
from dagster import Definitions
from dagster._config.structured_config import ConfigurableResource

class CredentialsResource(ConfigurableResource):
    username: str
    password: str

class FileStoreBucket(ConfigurableResource):

>     credentials: CredentialsResource

Reviewer 2: Thoughts on adding a code comment here to call this out? (cc @schrockn, if you have thoughts on this approach)

Reviewer: To call what out? I might be missing some context here. In general a big fan in comments in code samples as I think a lot of readers skip prose and just look at examples.

    region: str

    def write(self, data: str):
        get_filestore_client(
            username=self.credentials.username,
            password=self.credentials.password,
            region=self.region,
        ).write(data)

credentials = CredentialsResource(username="my_user", password="my_password")
defs = Definitions(
    assets=...,
    resources={
        "bucket": FileStoreBucket(
            credentials=credentials,
            region="us-east-1",
        ),
    },
)
```

> If we instead would like the configuration for our credentials to be provided at runtime, we can use the `configure_at_launch()` method to defer the construction of the `CredentialsResource` until runtime.

Reviewer 2: ```suggestion
If instead we want the configuration for our credentials to be provided at runtime, we can use the `configure_at_launch()` method to defer the construction of the `CredentialsResource` until runtime.
```

> Because `credentials` requires runtime configuration through the launchpad, it must also be passed to the `Definitions` object, so that configuration can be provided at runtime. Nested resources only need to be passed to the `Definitions` object if they require runtime configuration.

Reviewer 2: I love that this is so straightforward!

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_dep_job_runtime endbefore=end_new_resource_dep_job_runtime
credentials = CredentialsResource.configure_at_launch()

defs = Definitions(
    assets=...,
    resources={
        "credentials": credentials,
        "bucket": FileStoreBucket(
            credentials=credentials,
            region="us-east-1",
        ),
    },
)
```

> ---

Reviewer 2: [1]

### Testing configurable resources

You can test the initialization of a <PyObject object="ConfigurableResource"/> by constructing it manually. In most cases, you can construct your resource directly:

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_testing endbefore=end_new_resource_testing
from dagster._config.structured_config import ConfigurableResource, ConfigurableResourceAdapter

V = TypeVar("V")

class ResourceDependency(Generic[V]):
    def __get__(self, obj: "Resource", __owner: Any) -> V:
        return getattr(obj, self._name)

> 

Reviewer: I don't think we need this in the code example

class MyResource(ConfigurableResource):
    value: str

    def get_value(self) -> str:
        return self.value

def test_my_resource():
    assert MyResource(value="foo").get_value() == "foo"
```

If your resource requires other resources, then you can pass them as constructor arguments.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_testing_with_context endbefore=end_new_resource_testing_with_context
from dagster._config.structured_config import ConfigurableResource

class StringHolderResource(ConfigurableResource):
    value: str

class MyResourceRequiresAnother(ConfigurableResource):
    foo: StringHolderResource
    bar: str

def test_my_resource_with_context():
    string_holder = StringHolderResource(value="foo")
    resource = MyResourceRequiresAnother(foo=string_holder, bar="bar")
    assert resource.foo.value == "foo"
    assert resource.bar == "bar"
```

---

> ## Defining Pythonic IO managers

Reviewer 2: ```suggestion
## Defining Pythonic I/O managers
```

> Pythonic [IO managers](/concepts/io-management/io-managers) are defined as subclasses of <PyObject object="ConfigurableIOManager"/>, and similarly to Pythonic resources specify any configuration fields as attributes. Each subclass must implement a `handle_output` and `load_input` method, which are called by the Dagster runtime to handle the storing and loading of data.

Reviewer 2: ```suggestion
Pythonic [I/O managers](/concepts/io-management/io-managers) are defined as subclasses of <PyObject object="ConfigurableIOManager"/>, and similarly to Pythonic resources specify any configuration fields as attributes. Each subclass must implement a `handle_output` and `load_input` method, which are called by the Dagster runtime to handle the storing and loading of data.
```

Also:

> ... by the Dagster runtime to handle the storing and loading of data.

Is `the Dagster runtime` a thing? (I'm not trying to sound snarky, I honestly don't know - it read a bit like a typo)

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_io_manager endbefore=end_new_io_manager
class MyIOManager(ConfigurableIOManager):
    root_path: str

    def _get_path(self, asset_key: AssetKey) -> str:
        return self.root_path + "/".join(asset_key.path)

    def handle_output(self, context: OutputContext, obj):
        write_csv(self._get_path(context.asset_key), obj)

    def load_input(self, context: InputContext):
        return read_csv(self._get_path(context.asset_key))

defs = Definitions(
    assets=...,
    resources={
        "io_manager": MyIOManager(root_path="/tmp/")
    },
)

> ```

Reviewer: IMO after this section we should introduce the notion of the factory as I think it will be fairly common to do this for I/O managers *and* it is necessary to understand the factory in order to adapt the old-style resources

---

## Advanced resource patterns

The following sections describe more advanced patterns for using Pythonic resources.

### Adapting function-style resources

> In codebases which utilize [function-style](https://docs.dagster.io/concepts/resources#resources) resources with the <PyObject object="@resource"/> decorator, you can use the <PyObject object="ConfigurableResourceAdapter"/> class to adapt these resources to the new Pythonic resource pattern.

Reviewer 2: ```suggestion
In codebases that utilize [function-style](https://docs.dagster.io/concepts/resources#resources) resources with the <PyObject object="@resource"/> decorator, you can use the <PyObject object="ConfigurableResourceAdapter"/> class to adapt these resources to the new Pythonic resource pattern.
```

> The adapter subclass should list all config fields as attributes, and should implement a `wrapped_resource` property which returns the underlying resource definition to adapt.

Reviewer 2: ```suggestion
The adapter subclass should list all config fields as attributes and implement a `wrapped_resource` property which returns the underlying resource definition to adapt.
```

    prefix = context.resource_config["prefix"]
    return Writer(prefix)

# New adapter layer
class WriterResource(ConfigurableResourceAdapter):
    prefix: str

    @property
    def wrapped_resource(self) -> ResourceDefinition:
        return writer_resource

@asset
def my_asset(writer: Writer):
    writer.output("hello, world!")

defs = Definitions(assets=[my_asset], resources={"writer": WriterResource(prefix="greeting: ")})
```

### Using bare Python objects as resources

In some cases, you may want to use a bare Python object as a resource which is not a subclass of <PyObject object="ConfigurableResource"/>.

For example, you may want to directly pass a third-party API client into your assets or ops. This follows a similar pattern to using a <PyObject object="ConfigurableResource"/> subclass, however assets and ops which use these resources must [annotate](https://docs.python.org/3/library/typing.html#typing.Annotated) them with <PyObject object="Resource"/>.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_raw_github_resource endbefore=end_raw_github_resource
@asset
def public_github_repos(github: Resource[GitHub]):

>     return github.organization("dagster-io").repositories()

Reviewer: Recommend adding comment here to emphasize how annotations work
e.g.

`# Resource[Github] is treated exactly like Github for type checking purposes and the runtime type of the github parameter is GitHub. The purpose of Resource is to tell dagster that github is a resource rather than an upstream asset.`

defs = Definitions(
    assets=[public_github_repos],
    resources={"github": GitHub()},
)
```

> If you would like a Pythonic resource to depend on a bare Python object, you can use the <PyObject object="ResourceDependency"/> annotation to annotate the attribute in question. This indicates to Dagster that the attribute should be treated as a resource dependency.

Reviewer: When reading this in documentation form, it's a bit confusing. 

* What happens if you annotate it with just `Engine`. Is that an error? Why?
* What are the exact semantics of a resource dependency? Does it get an entry in the UI?

)
```

---

## Implementation details

### Resource input resolution

Resources defined by subclassing <PyObject object="ConfigurableResource"/> can take various unresolved inputs, including `EnvVar`s instead of string values or unconfigured resources returned by `configure_at_launch()` in place of configured resources.

Between construction time and runtime, the <PyObject object="ConfigurableResource"/> class resolves all of its attributes to their final values, meaning that when an attribute of a resource is accessed in the body of an asset or op, it will always be a fully resolved value.

In the following example, the `FileStoreBucket` constructor will accept the still-unconfigured `unconfigured_credentials_resource`, which will be resolved once the resource is configured at runtime.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_impl_details_resolve endbefore=end_impl_details_resolve
class CredentialsResource(ConfigurableResource):
    username: str
    password: str

class FileStoreBucket(ConfigurableResource):
    credentials: CredentialsResource
    region: str

    def write(self, data: str):

>         # In this context, `self.credentials` is ensured to

Reviewer: "In this context, `self.credentials` is ensured to be a CredentialsResource with valid values for `username` and `password`"

from dagster import Definitions, asset

class DatabaseResource(ConfigurableResource):
    table: str

    def read(self):
        ...

@asset
def data_from_database(db_conn: DatabaseResource):
    return db_conn.read()

defs = Definitions(
    assets=[data_from_database],
    resources={"db_conn": DatabaseResource.configure_at_launch()},
)
```

#### Configuring a resource at runtime in Python code

Then, configuration for the resource can be provided at runtime in the launchpad or in Python code using the `config` parameter of the <PyObject object="RunRequest" />:

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_runtime_launch endbefore=end_new_resource_runtime_launch
from dagster import sensor, define_asset_job, RunRequest

update_data_job = define_asset_job(name="update_data_job", selection=[data_from_database])

@sensor(job=update_data_job)
def table_update_sensor():
    tables = ...
    for table_name in tables:
        yield RunRequest(
            run_config=RunConfig(
                resources={
                    "db_conn": DatabaseResource(table=table_name),
                },
            ),
        )

> ```

Reviewer: Wanted to flag this this will potentially be confusing because the behavior here will be slightly different than passing resources to `Definitions`, in particular when raw resource objects are in play. I anticipate that people will copy and paste resource dictionaries around alternatively pass them to Definitions and this. There will be slight differences in behavior.

    writer.output("hello, world!")

defs = Definitions(assets=[my_asset], resources={"writer": WriterResource(prefix="greeting: ")})
```

### Adapting function-style I/O managers

Similar to with resources, in codebases that utilize [function-style](https://docs.dagster.io/concepts/io-management/io-managers) I/O managers with the <PyObject object="@io_manager"/> decorator, you can use the <PyObject object="ConfigurableIOManagerAdapter"/> class to adapt them to the new Pythonic I/O manager pattern.

The adapter subclass should list all config fields as attributes and implement a `wrapped_io_manager` property which returns the underlying I/O manager definition to adapt.

```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_io_adapter endbefore=end_io_adapter
from dagster import (
    Definitions,
    IOManagerDefinition,
    io_manager,
    IOManager,
    InputContext,
    OutputContext,
)
from dagster._config.structured_config import (
    StructuredIOManagerAdapter as ConfigurableIOManagerAdapter,
)
import os

> # Old code, cannot be changed for back-compat purposes

Reviewer: I think one modification which should make here is that that we can't change the public interface for the code, rather than not change it at all.

class OldFileIOManager(IOManager):
    def __init__(self, base_path: str):
        self.base_path = base_path

    def handle_output(self, context: OutputContext, obj):
        with open(os.path.join(self.base_path, context.step_key, context.name), "w") as fd:
            fd.write(obj)

    def load_input(self, context: InputContext):
        with open(
            os.path.join(
                self.base_path, context.upstream_output.step_key, context.upstream_output.name
            ),
            "r",
        ) as fd:
            return fd.read()

> @io_manager(config_schema={"base_path": str})

Reviewer: If we are allow to change code but keep interface, we can develop a pattern here where we can canonicalize the config schema. We define the config schema in the Pythonic style only and programmatically generate the old school schema. I think this is the pattern we should use when we start to do integration conversion. I don't think we need to block this guide on that but you should put it on your radar.
